man/Vector-comparison.Rd

```diff
@@ -319,7 +366,7 @@ setMethod(">", c("Vector", "Vector"),
           it works out-of-the-box on a \link{Vector} object for which those
           things work the expected way.
 
-    \item The \pkg{S4Vectors} package provides no \code{match} or \code{order}
+    \item The \pkg{S4Vectors} package provides no \code{order}
```

Comment:
Should we mention that `base::order()` delegates to `xtfrm()`? So `order()` will work as long as there is an explicit `xtfrm()`. Perhaps `xtfrm,Vector()` should check for the existence of an `order()` method before ending up in an infinite recursion?

Reply:
Regarding the check: I presume this would involve putting `hasMethod` for `order` in the `xtfrm` definition for `Vector`, and `stop`ping if nothing is found? I guess it couldn't hurt, provided the `hasMethod` call is cheap. We would get maximum value for money by using `selectMethod` with `optional=TRUE` to avoid having to do the method search twice.

Regarding the docs: might as well, though I doubt many people would know enough and have the need to take to take advantage of the differences between `xtfrm` and `order`.

Reply:
Just tried to add:

```r
    FUN <- selectMethod("order", class(x), optional=TRUE)
    if (is.null(FUN)) {
        # Avoid infinite recursion.
        stop(sprintf("order() is not implemented for %s objects", class(x)))
    }
```

to the `xtfrm,Vector-method` definition. Problem is that, if `order()` hasn't been implemented for a `Vector` class, `selectMethod` happily finds `base::order` and never enters the error condition.

`getMethod()` does the right thing but doesn't allow inheritance - I presume this could be problematic for anyone calling `xtfrm()` on a subclass of a class that has implemented `order()`.

I don't know what the solution might be here, but if there isn't one, the recursion hits the stack limit pretty quickly, so it's a relatively painless way of prodding the developer to implement `order()`.

Reply:
My suggestion on the docs is only for the developers, so that they realize that they could implement `xtfrm()` without implementing `order()`, if so desired. For the dispatch check, it's more about giving an informative error message. The assumption of course is that calling `base::order()` will result in the same `xtfrm()` method being called (technically inheritance could interfere). We could skip it, but one idea is to just compare the return value of `selectMethod()` to `base::order()` with `identical()`.

Reply:
```r
    FUN <- selectMethod("order", class(x), optional=TRUE)
    if (is.null(FUN) || identical(FUN, base::order)) {
        # Avoid infinite recursion.
        stop(sprintf("order() is not implemented for %s objects", class(x)))
    }
```
... is not much better, unfortunately, because `FUN` seems to be a weird `derivedDefaultMethod`. I can make it work via `identical(FUN@.Data, base::order)`... but it's not pretty.

Anyway, the docs are amended as suggested. I don't think even developers would even consider implementing `xtfrm` - certainly it's not a function that I had even heard about until now - but it's worth mentioning as they'll get stack limit errors if they _forgot_ to implement `order`.